### PR TITLE
NullRestricted attribute field class checks

### PR DESCRIPTION
**Pull request by theresa-m**
- NullRestricted fields must be a value type with an ImplicitCreation attribute that has the ACC_DEFAULT flag set. For non-static fields this should be checked during class creation and for static fields during the preparation stage of linking.
- This pr also adds static NullRestricted fields to the flattenedClassCache to make the described failure possible. Changes to allow non-static NullRestricted fields to be flattened will come in a later pull request.
- Move ImplicitCreation flag declarations because flags are no longer part of extramodifiers, and correct default flag value

Related: https://github.com/eclipse-openj9/openj9/issues/17340

### CODE REVIEW

**Comment by hangshao0** [comment]
There are merge conflicts that need to be resolved.

File: runtime/vm/createramclass.cpp [hunk 1]
```diff
@@ -2064,6 +2064,24 @@ checkFlattenableFieldValueClasses(J9VMThread *currentThread, J9ClassLoader *clas
 		const U_32 modifiers = field->modifiers;
 		J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(field);
 		U_8 *signatureChars = J9UTF8_DATA(signature);
+
+		if (J9_ARE_ALL_BITS_SET(modifiers, J9FieldFlagIsNullRestricted)) {
```
Comment (hangshao0) [inline runtime/vm/createramclass.cpp:2068]: During class loading, we should check non-static null-restricted fields and throw `IncompatibleClassChangeError` when necessary. This check can be moved into `if (J9_ARE_NO_BITS_SET(modifiers, J9AccStatic))` a few lines below. Probably you can do something like:

```
if (J9_ARE_NO_BITS_SET(modifiers, J9AccStatic)) {
	if (('Q' == signatureChars[0])
             || J9_ARE_ALL_BITS_SET(modifiers, J9FieldFlagIsNullRestricted)
        ) {
```

For static null-restricted fields, the `IncompatibleClassChangeError` should be thrown during preparation. The checks should be added inside:
https://github.com/eclipse-openj9/openj9/blob/080f0db81157518a64831ee0c7b55ffeb6547d62/runtime/vm/ClassInitialization.cpp#L591-L643

**Comment by theresa-m** [comment]
@hangshao0 I'm planning to add changes to flatten non-static NullRestricted field classes in another pull request since its not needed for these checks.

File: runtime/oti/j9javaaccessflags.h [hunk 2]
```diff
@@ -167,4 +165,8 @@
 #define J9StaticFieldRefFlagBits 0x3F
 #endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
 
+/* ImplicitCreation attribute flags */
+#define J9AccImplicitCreateHasDefaultValue 0x01
+#define J9AccImplicitCreateNonAtomic 0x20
```
Comment (hangshao0) [inline runtime/oti/j9javaaccessflags.h:170]: Is this intentional ? `J9AccImplicitCreateHasDefaultValue` is `0x1`, and `J9AccImplicitCreateHasDefaultValue` is `0x20`.

File: runtime/vm/createramclass.cpp [hunk 3]
```diff
@@ -2081,6 +2083,17 @@ checkFlattenableFieldValueClasses(J9VMThread *currentThread, J9ClassLoader *clas
 					result = FALSE;
 					break;
 				}
+
+				/* A NullRestricted field must be in a value class with an
+				 * ImplicitCreation attribute. The attribute must have the ACC_DEFAULT flag set.
+				 */
```
Comment (hangshao0) [inline runtime/vm/createramclass.cpp:2089]: Can you add a comment saying that the same check for static field is done in class preparation ?

File: runtime/oti/j9javaaccessflags.h (hunk 2 above)
Comment (theresa-m) [reply runtime/oti/j9javaaccessflags.h:170]: Thanks for the catch. default was also previously wrong and I corrected it here. I'll update nonatomic as well.

**Review by hangshao0 — APPROVED** [review]
LGTM.

**Comment by hangshao0** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by hangshao0** [comment]
Jenkins test sanity,extended winval jdknext

File: runtime/vm/createramclass.cpp [hunk 4]
```diff
@@ -2081,6 +2083,20 @@ checkFlattenableFieldValueClasses(J9VMThread *currentThread, J9ClassLoader *clas
 					result = FALSE;
 					break;
 				}
+
+				/* A NullRestricted field must be in a value class with an
+				 * ImplicitCreation attribute. The attribute must have the ACC_DEFAULT flag set.
+				 * Static fields will be checked during class preparation.
+				 */
+				if (J9_ARE_ALL_BITS_SET(modifiers, J9FieldFlagIsNullRestricted)) {
```
Comment (theresa-m) [inline runtime/vm/createramclass.cpp:2091]: Forgot this important line here :) the functional/Valhalla tests are passing for me locally now.

**Comment by hangshao0** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by hangshao0** [comment]
Jenkins test sanity,extended winval jdknext

**Pull request merged**